### PR TITLE
ref(grouping): Refactor matchers such that extracted frame comes from indiv. matchers

### PR DESCRIPTION
**Pull request by untitaker**
Significantly refactor interface such that globbing happens in one place (exactly one impl of `matches_frame`) and can be ported to Rust as one unit. Before we do that we need to invoke matches_frame in batches though

### CODE REVIEW

**Comment by jjbayer** [comment]
If we merge this, grouping will be slower again for a while, right? ~2.5 times slower according to test_benchmark.py for newstyle. But I guess that's OK as long as hierarchical grouping hasn't been rolled out?

File: src/sentry/grouping/enhancer/matchers.py [hunk 1]
```diff
@@ -59,34 +61,60 @@ def _get_function_name(frame_data: dict, platform: Optional[str]):
...
 
-    return match_frame
 
-
-class Match:
+class Match(ABC):
     description = None
 
+    @abstractmethod
+    def extract_frame_data(self, frames, idx, platform, exception_data):
```
Comment (jjbayer) [inline src/sentry/grouping/enhancer/matchers.py:81]: Extracting frame data per matcher means potentially executing expensive functions (e.g. `get_function_name_for_frame`) more than once per frame. Will this be compensated by the move to rust?

File: src/sentry/grouping/enhancer/matchers.py [hunk 2]
```diff
@@ -59,34 +61,60 @@ def _get_function_name(frame_data: dict, platform: Optional[str]):
...
+        elif config.required:
+            return False
+        else:
+            val = None
+
+        if val is None:
+            is_match = False
+        elif config.pattern_is_glob:
+            if config.path_like:
+                is_match = any(path_like_match(val, pat.encode("utf8")) for pat in config.patterns)
```
Comment (jjbayer) [inline src/sentry/grouping/enhancer/matchers.py:102]: Wouldn't it still be better to encode the pattern in the matcher's constructor, so it doesn't have to be repeated on every frame?

**Comment by untitaker** [comment]
@jjbayer Fair enough. I'll convert this into a draft and continue working on it without merging then. I just thought this would be a good unit to review on its own but perhaps that's not deployable.

**Pull request closed without merging**